Salty Bone Skunk

Medium

# Unsafe typecasting of `nounId` from `uint256` to `uint96` in `_createAuction()` allows  anyone to pause the `AuctionHouse` contract

### Summary

Unsafe typecasting of `nounId` from `uint256` to `uint96` in [`_createAuction()`](https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/NounsAuctionHouseV3.sol#L299) function leads to the following two vulnerabilities:-

1. Creating an NFT auction for an already ended and settled auction cause of a `nounId` reuse due to a silent integer overflow.
2. Anyone can pause the `NounsAuctionHouse`.

### Root Cause

The `_createAuction()` function is meant to create an auction. This function explicitly returns a `uint256`, i.e

```solidity
    /**
     * @notice Create an auction.
     * @dev Store the auction details in the `auction` state variable and emit an AuctionCreated event.
     * If the mint reverts, the minter was updated without pausing this contract first. To remedy this,
     * catch the revert and pause this contract.
     */
    function _createAuction() internal {
        try nouns.mint() returns (uint256 nounId) {     /// @audit-tag return uint256
            uint40 startTime = uint40(block.timestamp);
            uint40 endTime = startTime + uint40(duration);

            auctionStorage = AuctionV2({
                nounId: uint96(nounId),             /// @audit-tag unsafe cast from uint256 to uint96
                clientId: 0,
                amount: 0,
                startTime: startTime,
                endTime: endTime,
                bidder: payable(0),
                settled: false
            });

            emit AuctionCreated(nounId, startTime, endTime);
        } catch Error(string memory) {
            _pause();
        }
    }


```

The problem is that the `uint256 nounId` is then cast to a `uint96` as seen above. This is a problem when the `nounId` to be minted is greater than `type(uint96).max`.

Take for example the `nounId` to be minted is `type(uint96).max + 1`. When this `nounId` integer reaches the `_createAuction()` function and unsafely typecasted to `uint96`, the `nounId` that will be attempted to be minted will be `nounId` 0 as the typecasting silently overflows and wraps up to start counting from 0.

Take the example below to confirm that the overflow happens even when the Solidity version used is >0.8.0

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

contract Test {
    // type(uint256).max = 115792089237316195423570985008687907853269984665640564039457584007913129639935
    // type(uint96).max = 79228162514264337593543950335
    function functionA(uint256 _number) external pure returns (uint256) {
        return uint96(_number);
    }
}
```
Running it in remix produces the output below:-

when `_number` is 79228162514264337593543950336, i.e `type(uint96).max + 1` what is returned is 0

However, the issue at hand isn't as straightforward as it seems from above because of a small caveat, the try...catch block. When we try to mint nounId `type(uint96).max + 1`, we enter the try block which calls the `mint()` function from the [NounsToken](https://etherscan.io/token/0x9C8fF314C9Bc7F6e59A9d9225Fb22946427eDC03#code) which looks like the below:-

```solidity
    /**
     * @notice Mint a Noun to the minter, along with a possible nounders reward
     * Noun. Nounders reward Nouns are minted every 10 Nouns, starting at 0,
     * until 183 nounder Nouns have been minted (5 years w/ 24 hour auctions).
     * @dev Call _mintTo with the to address(es).
     */
    function mint() public override onlyMinter returns (uint256) {
        if (_currentNounId <= 1820 && _currentNounId % 10 == 0) {
            _mintTo(noundersDAO, _currentNounId++);
        }
        return _mintTo(minter, _currentNounId++);
    }
```

which in turn calls function `_mintTo()` which looks like the below:-

```solidity
    /**
     * @notice Mint a Noun with `nounId` to the provided `to` address.
     */
    function _mintTo(address to, uint256 nounId) internal returns (uint256) {
        INounsSeeder.Seed memory seed = seeds[nounId] = seeder.generateSeed(nounId, descriptor);

        _mint(owner(), to, nounId);
        emit NounCreated(nounId, seed);

        return nounId;
    }
```

This also in turn calls the `_mint()` function from OpenZeppelin which looks like the below:-

```solidity
/**
     * @dev Mints `tokenId`, transfers it to `to`, and emits two log events -
     * 1. Credits the `creator` with the mint.
     * 2. Shows transfer from the `creator` to `to`.
     *
     * WARNING: Usage of this method is discouraged, use {_safeMint} whenever possible
     *
     * Requirements:
     *
     * - `tokenId` must not exist.
     * - `to` cannot be the zero address.
     *
     * Emits a {Transfer} event.
     */
    function _mint(
        address creator,
        address to,
        uint256 tokenId
    ) internal virtual {
        require(to != address(0), 'ERC721: mint to the zero address');
        require(!_exists(tokenId), 'ERC721: token already minted');

        _beforeTokenTransfer(address(0), to, tokenId);

        _balances[to] += 1;
        _owners[tokenId] = to;

        emit Transfer(address(0), creator, tokenId);
        emit Transfer(creator, to, tokenId);
    }
```

OpenZeppelin's `mint()` checks that that tokenId to be minted isn't already minted, if the tokenId is minted, it reverts, i.e

```solidity
require(!_exists(tokenId), 'ERC721: token already minted');
```

With this, it means that when we attempt to mint token `type(uint96).max + 1`, OpenZeppelin's `mint()` function will revert as token Id 0 is already in existence. It reverts cause of the require statement

```solidity
require(!_exists(tokenId), 'ERC721: token already minted');
```

This revert is propagated up the call stack and lands us in the catch block of the function `_createAuction()`.

Note that the catch block

```solidity
catch Error(string memory) {
    _pause();
}
```

will handle all reverts with a reason string. That means, a `require(false, "reason")` error will be caught here. This is because this error returns the `Error(string)` error when triggered. See the [official Solidity documentation](https://docs.soliditylang.org/en/latest/control-structures.html#panic-via-assert-and-error-via-require)


This is exactly where we want to land as landing in this block will pause the contract.


If token Id 0(which belongs to the project's founders) is burnt, that will mean that we land in the try block and we end up kickstarting the auction of token Id 0. This is problematic considering what happens inside this try block, i.e 

```solidity
auctionStorage = AuctionV2({
    nounId: uint96(nounId),
    clientId: 0,
    amount: 0,
    startTime: startTime,
    endTime: endTime,
    bidder: payable(0),
    settled: false
    });
```

In our example, `nounId` 0's auction ended a very long time ago and has been settled. This translates to reassigning some very crucial metadata as seen above.

### Note

This bug is an edge case with no exploitable path as it leverages the fact that the attacker has to overflow `type(uint96).max` which is roughly 2.51 × 10^24 years. This translates to around 2.51 × 10^24 years worth of tokens. To put that into perspective, the current age of the universe is about 13.8 billion years. The universe would likely cease to exist long before reaching this number of tokens. The bug is more of a theoretical vulnerability than a realistic exploit. In essence, while the bug exists, it's so improbable. The practical likelihood of exploitation is essentially zero. While it highlights a precise technical issue, it has no real-world risk due to the astronomical scale required to trigger it.


### Internal pre-conditions

_No response_

### External pre-conditions

1. The protocol continues minting beyond `type(uint96).max` tokens. This isn't a problem considering that this is the main aim of the protocol, i.e
> One Noun, Every Day, Forever.

Here I'm assuming `type(uint96).max` < forever. I do acknowledge that this isn't practical as the universe would be long gone before this integer is even reached.

2. `nounId` 0 is burnt(if we want to kickstart its auction)

### Attack Path

1. A user waits until `type(uint96).max` NFTs have been minted.
2. The user then calls the external function [`NounsAuctionHouseV2::settleCurrentAndCreateNewAuction()`](https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/NounsAuctionHouseV2.sol#L99-L105) or [`NounsAuctionHouseV3::settleCurrentAndCreateNewAuction()`](https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/NounsAuctionHouseV3.sol#L115-L121) which in turn calls our problematic [`_createAuction()`](https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/NounsAuctionHouseV3.sol#L299) function.
3. The silent overflow happens and counting starts from 0.
4. If `nounId` 0 is burnt, the user will kickstart the noun's auction otherwise the user will pause the `AuctionHouse` contract.

### Impact

1. Malicious users can pause the `NounsAuctionHouseV3` contract as often as they like.
Of course, the DAO will still be able to unpause the contract by calling function [`unpause`](https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/NounsAuctionHouseV3.sol#L210-L221). But until it does so (the proposal would have to be first created and voted, which takes time), the contract will be paused and impossible to use.
2. A malicious user can kickstart the auction of an already ended auction.

### PoC

_No response_

### Mitigation

Don't typecast the `uint256` to `uint96`.